Visualize and understand GPU memory in PyTorch (#2560)
huggingface/blog: train_memory.md

Let's break down this graph into key parts:

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/train_memory/simple_profile_partitioned.png" width="1100" height="auto" alt="Simple profile partitioned">

1. **Model Creation**: Memory increases by 4 GB, corresponding to the model's size:

   \\( 10{,}000 \times 50{,}000 \text{ weights} + 50{,}000 \text{ biases in } \texttt{float32 }\text{(4 bytes)} \implies (5 \times 10^8) \times 4 \, \text{bytes} = 4 \, \text{GB}. \\)

   This memory (in blue) persists throughout execution.

2. **Input Tensor Creation (1st Loop)**: Memory increases by 200 MB matching the input tensor size:

   \\( 5{,}000 \times 10{,}000 \text{ elements in } \texttt{float32 }\text{(4 bytes)} \implies (5 \times 10^7) \times 4 \, \text{bytes} = 0.2 \, \text{GB}.\\)

3. **Forward Pass (1st Loop)**: Memory increases by 2 GB for the output tensor:  

   \\( 10{,}000 \times 50{,}000 \text{ elements in } \texttt{float32 }\text{(4 bytes)} \implies (5 \times 10^8) \times 4 \, \text{bytes} = 2 \, \text{GB}.\\)

> 4. **Input Tensor Creation (2nd Loop)**: Memory increases by 200 MB for a new input tensor. At this point, you might expect the input tensor from step 2 to be freed, but it isn't: the model retains its activation, so even if the tensor is no longer assigned to the variable `inputs`, it remains referenced by the model's forward pass computation. The model retains its activations because these tensors are required for the backpropagation process in neural networks. Try with `torch.no_grad()` to see the difference.

Reviewer: ```suggestion
4. **Input Tensor Creation (2nd Loop)**: Memory increases by 200 MB for a new input tensor. At this point, you might expect the input tensor from step 2 to be freed. Still, it isn't: the model retains its activation, so even if the tensor is no longer assigned to the variable `inputs`, it remains referenced by the model's forward pass computation. The model retains its activations because these tensors are required for the backpropagation process in neural networks. Try with `torch.no_grad()` to see the difference.
```

   Gradients are used to update the model’s parameters. Initially, the optimizer itself is initialized (green zone). This initialization is only done once. After that, the optimizer uses the gradients to update the model’s parameters. To update the parameters, the optimizer temporarily stores intermediate values (red zone). After the update, both the gradients (yellow) and the intermediate optimizer values (red) are discarded, freeing memory.

At this point, one training iteration is complete. The process repeats for the remaining iterations, producing the three memory spikes visible in the graph.  

Training profiles like this typically follow a consistent pattern, which makes them useful for estimating GPU memory requirements for a given model and training loop.

## 📐 Estimating Memory Requirements

From the above section, estimating GPU memory requirements seems simple. The total memory needed should correspond to the highest peak in the memory profile, which occurs during the **forward pass**. In that case, the memory requirement is (blue + greeen + orange):

\\( \text{Model Parameters} + \text{Optimizer State} + \text{Activations} \\)

> Is it that simple? Actually there is a trap. The profile can look different depending on the training setup. For example, reducing the batch size from 16 to 2 changes the picture:

Reviewer: ```suggestion
Is it that simple? Actually, there is a trap. The profile can look different depending on the training setup. For example, reducing the batch size from 16 to 2 changes the picture:
```

```diff
- inputs = torch.randint(0, 100, (16, 256), device="cuda")  # Dummy input
+ inputs = torch.randint(0, 100, (2, 256), device="cuda")  # Dummy input
```

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/train_memory/colorized_training_profile_2.png" width="1100" height="auto" alt="Colorized training profile 2">

Now, the highest peaks occur during the **optimizer step** rather than the forward pass. In this case, the memory requirement becomes (blue + green + yellow + red):

\\( \text{Model Parameters} + \text{Optimizer State} + \text{Gradients} + \text{Optimizer Intermediates} \\)

To generalize the memory estimation, we need to account for all possible peaks, regardless of whether they occur during the forward pass or optimizer step.

\\( \text{Model Parameters} + \text{Optimizer State} + \max(\text{Gradients} + {\text{Optimizer Intermediates}, \text{Activations}}) \\)

> Now that we've the equation, let's see how to estimate each component.

Reviewer: ```suggestion
Now that we have the equation, let's see how to estimate each component.
```
